theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,162 @@
+import theano.tensor.nnet.ctc
+from .type import (GpuArrayType, gpu_context_type)
+from .elemwise import GpuDimShuffle
+from theano.gradient import grad_undefined
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+from theano.tensor.nnet.ctc import ctc_available
+
+import os
+import os.path
+from . import pygpu
+
+
+class GpuConnectionistTemporalClassification(gof.COp):
+    """
+    GPU wrapper for Baidu CTC loss function.
+
+    Parameters
+    ----------
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    """
+    __props__ = ('compute_grad',)
+
+    _cop_num_inputs = 3
+    _cop_num_outputs = 2
+
+    func_file = "./c_code/ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_gpu)"
+
+    params_type = gpu_context_type
+
+    def __init__(self, compute_grad=True):
+        if not ctc_available():
+            raise RuntimeError('Baidu CTC is not available and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+
+        self.compute_grad = compute_grad
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+
+    def c_lib_dirs(self):
+        assert ctc_available.path is not None
+        return [ctc_available.path]
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray/ext_cuda.h',
+                'gpuarray_helper.h', 'gpuarray/types.h', 'gpuarray_api.h',
+                'gpuarray/array.h', 'gpuarray/util.h', 'gpuarray/extension.h']
+
+    def get_params(self, node):
+        return node.inputs[0].type.context
+
+    def make_node(self, activations, labels, input_lengths):
+        context_name = infer_context_name(activations)
+        t_activations = as_gpuarray_variable(activations,
+                                             context_name=context_name)
+        # Ensure activations array is C-contiguous
+        t_activations = gpu_contiguous(t_activations)
+
+        # Labels and input lengths are always on the CPU
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('Activations must use the float32 type!')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('Labels must use the int32 type!')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
```

Comment:
**NB**: sorry, I forgot, but check here that inputs have expected nim (3 for activations, 2 for labels, 1 for inputs_lengths).